File: app/src/main/java/com/handstandsam/maintainableespresso/features/home/HomePresenter.java
package com.handstandsam.maintainableespresso.features.home;

import android.content.Context;
import android.content.Intent;
import android.support.annotation.NonNull;

import com.google.gson.GsonBuilder;
import com.handstandsam.maintainableespresso.MyAbstractApplication;
import com.handstandsam.maintainableespresso.models.Category;
import com.handstandsam.maintainableespresso.models.User;
import com.handstandsam.maintainableespresso.repository.CategoryRepository;
import com.handstandsam.maintainableespresso.repository.SessionManager;

import java.util.List;

import javax.inject.Inject;

import timber.log.Timber;

public class HomePresenter {

    private final Context applicationContext;
    private final HomeActivity.HomeView view;

    @Inject
    SessionManager sessionManager;

    @Inject
    CategoryRepository categoryRepository;

    public HomePresenter(@NonNull HomeActivity.HomeView homeView) {
        this.view = homeView;
        this.applicationContext = homeView.getContext().getApplicationContext();
        ((MyAbstractApplication) applicationContext).getAppComponent().inject(this);
        Timber.d("HomePresenter: " + new GsonBuilder().create().toJson(categoryRepository));
    }

    public void onResume(Intent intent) {
        List<Category> categories = categoryRepository.getCategories();
        view.showCategories(categories);

        User currentUser = sessionManager.getCurrentUser();
        String welcomeStr = "Welcome back " + currentUser.getFirstname() + " " + currentUser.getLastname();
        view.setWelcomeMessage(welcomeStr);
    }
}

File: app/src/main/java/com/handstandsam/shoppingapp/features/home/HomePresenter.kt
package com.handstandsam.shoppingapp.features.home

import android.content.Intent
import com.handstandsam.shoppingapp.repository.CategoryRepo
import com.handstandsam.shoppingapp.repository.NetworkResult
import com.handstandsam.shoppingapp.repository.SessionManager
import com.handstandsam.shoppingapp.utils.exhaustive
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import timber.log.Timber

class HomePresenter(
    private val view: HomeActivity.HomePresenter,
    private val sessionManager: SessionManager,
    private val categoryRepo: CategoryRepo,
    scope: CoroutineScope
) : CoroutineScope by scope {

    fun onResume(intent: Intent) {
        launch {
            val categoriesResult = withContext(Dispatchers.Default) {
                categoryRepo.getCategories()
            }
            when (categoriesResult) {
                is NetworkResult.Success -> {
                    view.showCategories(categoriesResult.body)
                }
                is NetworkResult.Failure -> {
                    Timber.d("onError")
                }
            }.exhaustive
        }


        val currentUser = sessionManager.currentUser
        val welcomeStr = "Welcome back " + currentUser?.firstname + " " + currentUser?.lastname
        view.setWelcomeMessage(welcomeStr)
    }
}
